Health Report UX improvements

peterflynn opened the pull request:
This is a superset of #10887.  @prksingh sorry, I hope this doesn't step on anything you'd already started writing... I set out to just tweak the strings & CSS, and this sort of snowballed after that :-P  I tried to split this cleanly into separate commits, so if you really wanted to take only the visual changes I think you could do that.

Here are the differences from 10887:
- Don't send data 30 until minutes after showing the popup, regardless of how quickly the user closes it - this gives users time to read more and opt out if desired, even if they close the popup before changing the pref.
  - Note: I took this opportunity to simplify the send-time calculations a bit by changing viewstate from "last time sent" to "next to to send."
- Tighten wording in Health Data notification popup (ran this by @ryanstewart & @njx for feedback)
- Add similar wording for context to Health Data preferences/preview dialog
- Show full preferences dialog when Help menu item is invoked, even if user is currently opted out (it's actually impossible to opt back in if we showed the popup instead, as 10887 does)
- Improve popup layout & show/hide animation
- Add tooltips to links in popup & dialog
- Fix bug with popup placement on Linux (due to HTML menu bar)
- Remove unused `HealthDataNotification.showDialogHealthDataNotification()` and change unit test to check notification popup instead
- Small refactorings: rename ExtensionManager flag for clarity; use ExtensionManager constant instead of duplicating string literal; share link-tooltip code better.

prksingh committed:
Adding a new popup for the first healthDataNotification

prksingh committed:
Making some CSS changes and adding close button to the popup

prksingh committed:
Changing some UI for the first time health data popup

prksingh committed:
Update healthdata-notification-dialog.html

prksingh committed:
Fixing JSHint issues that are causing the travis build to fail,

removing some unnecessary CSS

peterflynn committed:
- Tighten wording in Health Data notification popup
- Add similar wording for context to Health Data preferences/preview dialog
- Show full preferences dialog when Help menu item is invoked, even if user
is currently opted out (it's actually impossible to opt back in if we showed
the popup instead)
- Improve layout & popup transition in/out
- Add tooltips to links in popup & dialog
- Fix bug with popup placement on Linux (HTML menu bar)
- Remove unused HealthDataNotification.showDialogHealthDataNotification()
and change unit test to check notification popup instead

peterflynn committed:
Don't sent Health Data 30 until minutes after showing the popup, regardless
of how quickly the user closes it - this gives users time to read more and
opt out if desired, even if they close the popup before changing the pref.

Simplifies send-time calculations a bit by changing viewstate from "last
time sent" to "next to to send."

peterflynn committed:
Rename ExtensionManager flag for clarity
Use ExtensionManager constant instead of duplicating string literal

peterflynn committed:
Merge remote-tracking branch 'origin/master' into pflynn/health-report-ux

* origin/master: (27 commits)
  - String tweaks for Mac 'command line shortcut' feature. - Ensure URLs shown in dialogs for this feature include a tooltip of the URL
  Incorporating some code review comments
  Remove extra line
  Fix an escaping
  Getting Started project: remove useless whitespace, fix an arrow
  Fix a string
  Fix case before semicolons
  Rename Health Data names
  Use “\u2013” instead of n dash (–)
  removing the folds key from prefernece file as it now resides in the state file.
  Stop using supplemental Unicode arrow in the code-folding placeholder - Windows may not have glyphs for it and simpler Unicode arrows don't look good. Use simple "..." text instead (vertically aligned so it's more like an icon than text).
  Finnish translation, release 1.3
  Fixing lint errors
  using PreferencesManager.stateManager to set default foldKeys for code folding.
  Update strings.js
  Update strings.js
  fix issue#10877  Code folding is always reset after closing and reopening file or project
  Incorporating some code review comments and string changes
  Moved the command line tool instal menu to File menu
  Fix stray tab chars Remove unused `#code-folding-settings-dialog` CSS rule Fix some comment whitespace
  ...

peterflynn committed:
Fix duplicated code for adding tooltips to links

Tweak vertical alignment of Health Data prefs checkbox

peterflynn commented:
Here's what the UI looks like now with this PR:

![screen shot 2015-04-15 at 4 40 58 pm](https://cloud.githubusercontent.com/assets/1197510/7171520/54295660-e38e-11e4-8a85-569730609b18.png)

![screen shot 2015-04-15 at 4 41 37 pm](https://cloud.githubusercontent.com/assets/1197510/7171521/542bd566-e38e-11e4-919d-65e17400ef4a.png)

peterflynn commented:
@prksingh et al: regardless of whether we merge this or some other variation of PR #10887, we should probably do another test pass on this feature overall since the notification & the data-send timing is changing significantly from how it used to behave.  (I tested this a bunch locally of course, but it's important to have a 2nd set of eyes).

prksingh commented:
This looks good @peterflynn Thanks for the UI tweaks! I will take another look at this later and merge this.

@prafulVaishnav Can you also have a look at the changes?

prafulVaishnav commented:
@peterflynn If user have opt-out then we should not be showing the health data preview as in that case we will not be logging/sending anything. Instead we should show the preference dialog where he can opt-in.

peterflynn commented:
@prafulVaishnav That change is by design -- if the user has opted out, it's still useful for them to be able to see what data _would_ be sent if they were to opt back in.  Maybe they'll decide it's not so scary after all, for example :-)  The UI is now unified: the dialog contains both the preference and the preview.

prafulVaishnav commented:
@peterflynn @nethip We had some discussion on it based on which we decided to keep different dialog box for preference(to opt-in) and preview(containing option to opt-out) Previewing the data on opt-out will give the feeling to the user that we are still tracking the data which we aren't. At some later stage, we will be logging more data like user events, workflows, etc.

peterflynn commented:
It seem important for people to be able to see the data and then decide if they're ok with sending it.  The Firefox one works this way too.  Given that the checkbox is right there, do you think people would really misunderstand?  @ryanstewart, thoughts?

nethip commented:
@peterflynn I think people would have more confidence when we show them a dialog saying we are not tracking your data anymore as you have opted out against showing them the data even they have opted out. Talking about the Firefox model, I personally would feel a little odd to see the tracking data though I have opted out of it.

peterflynn commented:
One of the points of the Firefox Health Report -- and hopefully here too -- is that the data can benefit the user directly by showing them stats about their performance, average project size, etc.  So you don't want to refuse the user access to their own data just because they've opted out of sharing it with others.  This is also useful for field diagnostics -- e.g. if a user complains Brackets is running slow, we could ask them to check the Health Report dialog and see if the projects they tend to edit are unusually large, etc.

What about having the string vary depending on opt-in state?  E.g. when opted out, it could say something like: "Below is a preview of the data that _would_ be sent in your next Health Report if enabled (sharing your Health Report is currently disabled)."

nethip commented:
@peterflynn Yeah I am OK with the changing of the wording. We need to clearly distinguish between diagnostic data and tracking data. I am in for showing the former in case I opt out. 

I can compare this to be a browser cookie just waiting to be used.

nethip commented:
That being said I am fine if we can express the same with good wording.

MarcelGerber commented on src/extensibility/ExtensionManager.js:
```diff
@@ -789,7 +789,7 @@ define(function (require, exports, module) {
     exports.getAvailableUpdates     = getAvailableUpdates;
     exports.cleanAvailableUpdates   = cleanAvailableUpdates;
-    exports.isRegistryObjectUpdated = false;
+    exports.hasDownloadedRegistry = false;
```
nit: indentation.
Also, a newline before this line would separate the functions above and this state variable even more.

MarcelGerber commented on src/widgets/Dialogs.js:
```diff
@@ -413,6 +413,26 @@ define(function (require, exports, module) {
+            var $elem = $(elem);
+            var url = $elem.attr("href");
+            if (url && url !== "#") {
+                $elem.attr("title", url);
```
Maybe you should make sure the element doesn't have a title yet.

peterflynn commented:
@nethip Ok, I'll push up the dynamic wording changes tonight.

prafulVaishnav commented on src/extensions/default/HealthData/HealthDataPopup.js:
```diff
@@ -0,0 +1,84 @@
+/*
```
Later we may want to generalize this for the popup other than HealthDataNotification.

peterflynn committed:
More Health Data UX tweaks, after code review:
- Clarify that preview of data it NOT getting sent if user opted out
- Simplify wording of opt in/out checkbox
- Dialogs.addLinkTooltips(): don't touch links that already have a tooltip
- Minor code style tweaks

peterflynn commented:
@nethip It felt weird having the wording jump around as the checkbox is toggled, so I went with a more static set of wording that emphasizes the previewed data isn't sent if you've opted out:

![screen shot 2015-04-16 at 11 49 25 pm](https://cloud.githubusercontent.com/assets/1197510/7197274/dd6325ae-e493-11e4-9165-f743a115a012.png)

(this also simplifies the checkbox label itself, after some discussion with NJ & @ryanstewart)

@MarcelGerber Thanks for the feedback -- changes pushed to fix those nits too

@prafulVaishnav I agree we may want a more generalized Brackets notification UI widget in the future.  Are you ok with waiting until a future PR to generalize this, though?  It's good to have more concrete use cases in mind before generalizing anyway...

prafulVaishnav commented:
@peterflynn Yes in future we will take care of generalizing it. Thanks a lot Peter. Merging this.
